src/test/fuzz/inputfetcher.cpp

```diff
@@ -0,0 +1,150 @@
+    {
+        abort();
+    }
+};
+
+FUZZ_TARGET(inputfetcher)
+{
+    SeedRandomStateForTest(SeedRand::ZEROS);
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+
+    const auto worker_threads{
+        fuzzed_data_provider.ConsumeIntegralInRange<int32_t>(2, 4)};
+    InputFetcher fetcher{static_cast<size_t>(worker_threads)};
+
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
+        CBlock block;
+        Txid prevhash{Txid::FromUint256(ConsumeUInt256(fuzzed_data_provider))};
+
+        DbMap db_map{};
+        std::map<const COutPoint, const Coin> cache_map{};
+
+        DbCoinsView db(db_map);
+
+        NoAccessCoinsView back;
+        CCoinsViewCache cache(&back);
+
+        LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
+            CMutableTransaction tx;
+
+            LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10) {
+                const auto txid{fuzzed_data_provider.ConsumeBool()
+                    ? Txid::FromUint256(ConsumeUInt256(fuzzed_data_provider))
+                    : prevhash};
+                const auto index{fuzzed_data_provider.ConsumeIntegral<uint32_t>()};
+                const COutPoint outpoint(txid, index);
+
+                tx.vin.emplace_back(outpoint);
+
+                std::optional<Coin> maybe_coin;
+                if (fuzzed_data_provider.ConsumeBool()) {
+                    Coin coin{};
+                    coin.fCoinBase = fuzzed_data_provider.ConsumeBool();
+                    coin.nHeight =
+                        fuzzed_data_provider.ConsumeIntegralInRange<int32_t>(
+                            0, std::numeric_limits<int32_t>::max());
+                    coin.out.nValue = ConsumeMoney(fuzzed_data_provider);
+                    maybe_coin = coin;
+                } else {
+                    maybe_coin = std::nullopt;
+                }
+                db_map.try_emplace(outpoint, std::make_pair(
+                    maybe_coin,
+                    fuzzed_data_provider.ConsumeBool()));
+
+                // Add the coin to the cache
+                if (fuzzed_data_provider.ConsumeBool()) {
+                    Coin coin{};
+                    coin.fCoinBase = fuzzed_data_provider.ConsumeBool();
+                    coin.nHeight =
+                        fuzzed_data_provider.ConsumeIntegralInRange<int32_t>(
+                            0, std::numeric_limits<int32_t>::max());
+                    coin.out.nValue =
+                        fuzzed_data_provider.ConsumeIntegralInRange<int64_t>(
+                            -1, MAX_MONEY);
+                    cache_map.try_emplace(outpoint, coin);
+                    cache.EmplaceCoinInternalDANGER(
+                        COutPoint(outpoint),
+                        std::move(coin));
+                }
+            }
+
+            prevhash = tx.GetHash();
+            block.vtx.push_back(MakeTransactionRef(tx));
+        }
+
+        fetcher.FetchInputs(cache, db, block);
```

Comment:
We shouldn't test this with invalid (empty) blocks:

```suggestion
        if (block.vtx.empty()) continue;
        fetcher.FetchInputs(cache, db, block);
```

Reply:
Why not? The InputFetcher should not make assumptions about the structure of the block being passed in.

Reply:
Yeah, InputFetcher doesn't know if a block is consensus valid or not yet. It hasn't passed `ConnectBlock` yet before entering here.